morgen: 3.0.1 -> 3.1.6

justanotherariel opened the pull request:
## Description of changes
* Implement an updateScript for the morgen package.
* Update morgen to version 3.1.6
* Add `meta.mainProgram = "morgen"`.

Solves #274201

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

WolfangAukang commented on pkgs/applications/office/morgen/default.nix:
```diff
@@ -3,11 +3,11 @@
-    url = "https://download.todesktop.com/210203cqcj00tw1/morgen-${version}.deb";
-    sha256 = "sha256-lj+V5mntZzED2ZS62Uwlt/vTXwSuwzXeuEw8y/bA6og=";
+    url = "https://download.todesktop.com/210203cqcj00tw1/morgen-3.1.3-build-231204j94ts21rg.deb";
+    sha256 = "sha256-R7yqELmnCz9v9xSuL/1rhTQpfax/PDV6CM4Q7bbsk7c=";
```
```suggestion
    hash = "sha256-R7yqELmnCz9v9xSuL/1rhTQpfax/PDV6CM4Q7bbsk7c=";
```

Small fix

WolfangAukang reviewed (changes requested):
These changes are more like an update the configuration. Can you please also add `meta.mainProgram = "morgen"`?

WolfangAukang commented:
Result of `nixpkgs-review pr 274212` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>morgen</li>
  </ul>
</details>

justanotherariel commented:
Added the requested changes @WolfangAukang

mdehollander commented:
Hi, I just opened a PR to adjust the download link, see https://github.com/NixOS/nixpkgs/pull/275269 I checked on Friday if there was some ongoing work, but then there wasn;t. Over the weekend the download link has been fixed by the todesktop.com devs.

The link structure can be `https://dl.todesktop.com/[[your_app_id]]/versions/[[version]]` see https://www.todesktop.com/docs/recipes/download-links-from-your-website#download-links, so for the 3.1.3 this is https://dl.todesktop.com/210203cqcj00tw1/versions/3.1.3/linux/deb. This does not include the build string, so I think this should be stable with the next update. And then the ryantm bot can update it automatically.

justanotherariel commented:
I was trying to find the link structure of todesktop but couldn't find it, so thank you, @mdehollander! I also thought it would be nice to have an update script for the Morgen app, so I spent a few hours trying to implement it. I tested it, and it should work (the update script did the final update to 3.1.3). It definitely was a steep learning curve, and the todesktop API is not the nicest (head requests return a 404, and I had to figure out how to get the request headers of a GET request without downloading the file.)  But all good now :) Feel free to give me feedback on my update script.

@WolfangAukang I think it's ready to merge now.

Janik-Haag commented on pkgs/applications/office/morgen/default.nix:
```diff
@@ -3,11 +3,11 @@
-    url = "https://download.todesktop.com/210203cqcj00tw1/morgen-${version}.deb";
-    sha256 = "sha256-lj+V5mntZzED2ZS62Uwlt/vTXwSuwzXeuEw8y/bA6og=";
+    url = "https://dl.todesktop.com/210203cqcj00tw1/versions/${version}/linux/deb";
+    hash = "sha256-AzH1rZFqEH8sovZZfJykvsEmCedEZWigQFHWHl6/PdE=";
```
there is a hash missmatch, probably because upstream changed :/

justanotherariel replied:
Sorry, I somehow messed that up while preparing commits. Can you try again. Should be fixed now.

justanotherariel committed:
morgen: add updateScript and mainProgram

justanotherariel committed:
morgen: 3.0.1 -> 3.1.6

justanotherariel committed:
maintainers: add justanotherariel

Janik-Haag reviewed (approved):
LGTM, but I will wait with merging for a few days so @WolfangAukang has a chance to also review the pr again.

justanotherariel commented:
I just saw that somehow I ended up with the old sha256 attribute again (which was the request of @WolfangAukang), so I fixed that (again). That is the only thing I changed in the latest push.

WolfangAukang commented:
Now I get the following issue:
```
error: builder for '/nix/store/ny9blrv7g24svfzjrjksarfl1w258wly-deb.drv' failed with exit code 1;
       last 10 log lines:
       > Warning: Problem : HTTP error. Will retry in 1 seconds. 3 retries left.
       >   0    16    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
       > curl: (22) The requested URL returned error: 500
       > Warning: Problem : HTTP error. Will retry in 2 seconds. 2 retries left.
       >   0    16    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 500
       > Warning: Problem : HTTP error. Will retry in 4 seconds. 1 retries left.
       >   0    16    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 500
       > error: cannot download deb from any mirror
       For full logs, run 'nix log /nix/store/ny9blrv7g24svfzjrjksarfl1w258wly-deb.drv'.
```

kilianar commented:
It appears that the source is currently unavailable at the provided URL. Upon querying https://dl.todesktop.com/210203cqcj00tw1/versions/3.1.3/linux/deb, the server responds with a status code 500, leading to the build failure mentioned by @WolfangAukang.

justanotherariel commented:
Ugh. Apparently, they deleted the old version 3.1.3 upon releasing 3.1.4. I will update to the current version. Thankfully, the nixos cache system should still make old versions downloadable... if I understood it correctly.

justanotherariel commented:
There seems to be an issue with their backend. I can download the current version with neither:
- Old Link: https://download.todesktop.com/210203cqcj00tw1/morgen-3.1.4.deb (Error: 404)
- New Link: https://dl.todesktop.com/210203cqcj00tw1/versions/3.1.4/linux/deb (Error: 500)

I can only download the newest version with:
https://dl.todesktop.com/210203cqcj00tw1/linux/deb/x64

How should we proceed in this case? Wait, until they have resolved the issue on their end? Or use the link which will always download the newest version - leading to hash mismatches from time to time. This should only be an issue for those that do not make use of the nixos cache though.

WolfangAukang commented:
@justanotherariel Can you try the URL (old & new) with the current version of Morgen on Nixpkgs? (On my copy it is 3.0.1) If it builds with no issues, then I think we can use the new URL with no issues.

Janik-Haag commented on maintainers/maintainer-list.nix:
```diff
@@ -9214,6 +9214,12 @@
     githubId = 5124422;
     name = "Julien Urraca";
   };
+  justanotherariel = {
+    email = "ariel@ebersberger.io";
+    github = "justanotherariel";
+    githubId = 31776703;
+    name = "Ariel Ebersberger";
+  };
```
I think I missed this the first time around, but this should go into it's own commit called: `maintainers: add justanotherariel`

justanotherariel replied:
Ok, I will fix it as soon as we have decided on what to do with the download link

Janik-Haag commented:
The url also stopped working for me :(

justanotherariel commented:
@WolfangAukang Yes, same behavior.

So, to summarize:
1. We can download morgen 3.0.1 with the old link structure (i.e. https://download.todesktop.com/210203cqcj00tw1/morgen-3.0.1.deb), but not with the new one: https://dl.todesktop.com/210203cqcj00tw1/versions/3.0.1/linux/deb (Error 500)
2. Neither version 3.1.3 nor 3.1.4 are downloadable by either link structure:
  a. 3.1.3 old: https://download.todesktop.com/210203cqcj00tw1/morgen-3.1.3.deb (Error 404)
  b. 3.1.3 new: https://dl.todesktop.com/210203cqcj00tw1/versions/3.1.3/linux/deb (Error 500)
  c. 3.1.4 old: https://download.todesktop.com/210203cqcj00tw1/morgen-3.1.4.deb (Error 404)
  d. 3.1.4 new: https://dl.todesktop.com/210203cqcj00tw1/versions/3.1.4/linux/deb (Error 500)

The only way to get the newest version (currently 3.1.4), is by using the rolling release link: https://dl.todesktop.com/210203cqcj00tw1/linux/deb/x64

Staying at 3.0.1 is not an option though, as it would require to stick to electron version 15 which has reached EOL and will thus receive no more security patches.

I would thus propose to use the rolling release version to update the package. This will produce hash mismatches should someone try to build from source (instead of using the nix cache) while the package in nixpkgs hasn't been updated. I think this is a fair trade-off, however, as the package, otherwise, is not usable without introducing a security hole. We should then continue to monitor the situation and update the package with the new link structure, as soon as the 500 error is resolved on their side.

kilianar commented:
I believe we could duplicate the present release into the web archive and employ it as our primary source. Doing so would grant us a stable URL, marked by a timestamp denoting the snapshot. This practice aligns with the approach adopted by numerous other packages in nixpkgs.

justanotherariel commented:
Oh, I didn't think of that! Sounds like a good solution. I will update the PR as soon as I get to it.

justanotherariel commented:
So, one more question. With the introduction of the waybackmachine mechanism, the update script I created is now broken, as it can't save the new morgen version in the waybackmachine automatically. I could now do one of the following: 
a. disable the updateScript but keep it in the repository until the 500 error is resolved and we can use the new link structure - as it will start working as soon as that happens.
b. update the updateScript to save the new version automatically in the waybackmachine (but I think I would need another package as a point of reference)
c. remove the updateScript entirely, though I would hate to let my efforts go to waste.

kilianar commented:
I searched through the packages in nixpkgs aiming to locate those that retrieve data from the web archive and possess an update script capable of automatically archiving the latest release to the web archive. Regrettably, I couldn't find any files meeting these criteria by grepping for files containing both `archive.org` and `updateScript`. 

Relying on a non-functional update script that might be functional again in the future isn't an ideal approach. I propose that, for the time being, we refrain from including an update script. Instead, we could consider incorporating it once a stable download URL becomes available again. If it appears that there won't be a stable download URL in the foreseeable future, we could also consider rewriting the script to handle automatic backups to the web archive, following your suggestion.

mdehollander commented:
>     1. We can download morgen 3.0.1 with the old link structure (i.e. https://download.todesktop.com/210203cqcj00tw1/morgen-3.0.1.deb), but not with the new one: https://dl.todesktop.com/210203cqcj00tw1/versions/3.0.1/linux/deb (Error 500)
> 
>     2. Neither version 3.1.3 nor 3.1.4 are downloadable by either link structure:
>        a. 3.1.3 old: https://download.todesktop.com/210203cqcj00tw1/morgen-3.1.3.deb (Error 404)
>        b. 3.1.3 new: https://dl.todesktop.com/210203cqcj00tw1/versions/3.1.3/linux/deb (Error 500)
>        c. 3.1.4 old: https://download.todesktop.com/210203cqcj00tw1/morgen-3.1.4.deb (Error 404)
>        d. 3.1.4 new: https://dl.todesktop.com/210203cqcj00tw1/versions/3.1.4/linux/deb (Error 500)
> 

I send David from ToDesktop a message and he fixed the new link structure (with version) in 1h or so. Above links with a 500 error do work now. Assuming this link is stable (or can be fixed quickly), this might help fix the packaging issues.

justanotherariel commented:
@mdehollander great! dumb luck that I was just about to push the version with the web archive workaround... oh well. What's one more time editing these commits at this point.
So, with the latest push:
- Fixed the commit naming convetion pointed out by @Janik-Haag
- Fixed the update script (to handle the new amd64 extension and make it less fragile.
- Upraded morgen to version 3.1.6 

Tested the package and update script, so I hope everything's fine.

Janik-Haag reviewed (approved):
LGTM
